debian/rtl_sdr.1
.TH "rtl_sdr" 1 "0.5.0" RTL-SDR "User Commands"
.SH NAME
rtl-sdr \- an I/Q recorder for RTL2832 based DVB-T receivers
.SH DESCRIPTION
Uses a re-purposed DVB-T receiver as a software defined
radio to receive signals in I/Q data form. Written for
and incorporated in the osmocom rtl-sdr project.
.LP
In-Phase and Quadrature Phase data can faithfully represent
all of the information in a band of frequencies centered
on a carrier signal frequency.
.LP
Much software is available for the RTL2832. Most of the user-level
packages rely on the librtlsdr library which comes as part of the
rtl-sdr codebase. This codebase contains both the library itself and
also a number of command line tools such as rtl_test, rtl_sdr,
rtl_tcp, and rtl_fm. These command line tools use the library to test
for the existence of RTL2832 devices and to perform basic data
transfer functions to and from the device.
.LP
Because most of the RTL2832 devices are connected using USB, the
librtlsdr library depends on the libusb library to communicate with
the device.
.SH USAGE
This program captures information from a band of frequencies
and outputs the data in a form useful to other software radio
programs.
.SH SYNOPSIS
.B  rtl_sdr [-f freq] [OPTIONS] [output file]
.SH OPTIONS
.IP "-f frequency_to_tune_to [Hz]"
.IP "-s samplerate (default: 2048000 Hz)"
.IP "-d device_index (default: 0)"
.IP "-g gain (default: 0 for auto)"
.IP "-p ppm_error (default: 0)"
.IP "-b output_block_size (default: 16 * 16384)"
.IP "-n number of samples to read (default: 0, infinite)"
.IP "-S force sync output (default: async)"
.IP tfilename
 (a '-' dumps samples to stdout)
.SH EXAMPLES
.IP "Example: To tune to 392.0 MHz, and set the sample-rate to 1.8 MS/s, use:"
 ./rtl_sdr /tmp/capture.bin -s 1.8e6 -f 392e6
.LP
to record samples to a file or to forward the data to a fifo.
.LP
If the device can't be opened, make sure you have the appropriate
rights to access the device (install udev-rules from the repository,
or run it as root).
.SH SEE ALSO
gnuradio(1)
.LP
RTL-SDR wiki documentation:
.B http://sdr.osmocom.org/trac/wiki/rtl-sdr
.LP
Other rtl-sdr programs:
.sp
rtl_adsb(1), rtl_eeprom(1), rtl_fm(1), rtl_power(1), rtl_tcp(1), rtl_test(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2013 A. Maitland Bottoms <bottoms@debian.org>
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


debian/changelog
rtl-sdr (2.0.1) unstable; urgency=medium

  * debian/changelog: update for 2.0.0 and 2.0.1
  * gitignore: add files created by autotools

 -- Oliver Smith <osmith@sysmocom.de>  Fri, 03 Nov 2023 10:19:41 +0100

rtl-sdr (2.0.0) unstable; urgency=medium

  [ Derrick Pallas ]
  * tuner_r82xx: fix short-write in r82xx_read

  [ Steve Markgraf ]
  * Add rtl_biast as install target

  [ Eric Wild ]
  * cmake: populate pkgconfig file with prefix
  * fix windows build

  [ David Neiss ]
  * rtl_tcp: Extracted some constants out of printf strings

  [ Oliver Smith ]
  * rtl_tcp: put new DEFAULT_* constants in defines

  [ Doug Hammond ]
  * rtl_fm: add a new option to select 2nd direct sampling mode

  [ Martin Hauke ]
  * Fix minGW build

  [ jvde.github ]
  * lib: force wait state after cancel of usb transfer

  [ Clayton Smith ]
  * lib: Stop applying workaround for libusb < 1.0.9

  [ Tobias Girstmair ]
  * Fix signal handler from getting stuck in an endless loop

  [ rtlsdrblog ]
  * add rtl-sdr blog v4 support
  * fix rtl_tcp on macos
  * fix rtl_tcp error on windows when hints not initialized to 0
  * add blog v4 upconverter gpio switch
  * add -D direct sampling flag to rtl_tcp
  * add direct sampling to rtl_sdr

  [ Steve Markgraf ]
  * change version to 2.0.0

 -- Oliver Smith <osmith@sysmocom.de>  Fri, 03 Nov 2023 10:15:31 +0100

rtl-sdr (0.6) unstable; urgency=medium

  * New upstream release
  * with modernize-cmake patch

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 06 Oct 2018 20:28:26 -0400

rtl-sdr (0.6~git) unstable; urgency=medium

  * New upstream release

 -- Harald Welte <laforge@gnumonks.org>  Sun, 06 Jun 2018 15:09:42 +0200

rtl-sdr (0.5.4-1) unstable; urgency=medium

  * New upstream release
  * update to v0.5.4-3-ga854ae8
    use USB zero-copy transfers if possible

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 12 May 2018 16:49:43 -0400

rtl-sdr (0.5.3-14) unstable; urgency=medium

  * update to v0.5.3-20-g4520f00 (Closes: #892974)
  * minimal ipv6 support (Closes: #870804)
  * VCS to salsa
  * AppStream metadata.xml

 -- A. Maitland Bottoms <bottoms@debian.org>  Mon, 16 Apr 2018 20:45:53 -0400

rtl-sdr (0.5.3-13) unstable; urgency=medium

  * build with libusb-1.0-0-dev stub on hurd-i386
  * initial ipv6 support for rtl_tcp...

 -- A. Maitland Bottoms <bottoms@debian.org>  Thu, 23 Nov 2017 15:59:40 -0500

rtl-sdr (0.5.3-12) unstable; urgency=medium

  * add new HanfTek dongle
  * Bias T support (Closes: #854378, #842249)

 -- A. Maitland Bottoms <bottoms@debian.org>  Wed, 23 Aug 2017 23:31:27 -0400

rtl-sdr (0.5.3-11) unstable; urgency=medium

  * correct invocation of rm_conffile (Thanks Chris!) (Closes: #838161)
  * drop uploaders line on advice of MIA team. (Closes: #836590)

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 08 Oct 2016 11:17:47 -0400

rtl-sdr (0.5.3-10) unstable; urgency=medium

  * remove rtl-sdr-blacklist.conf on upgrade. Thanks Bob! (Closes: #829517)

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 09 Jul 2016 23:38:24 -0400

rtl-sdr (0.5.3-9) unstable; urgency=medium

  * Edit of debian/librtlsdr0.udev in 0.5.3-8 was a no-op. Real fix
    done to debian/patches/use-udev-uaccess-rules. (Closes: #825073)

 -- A. Maitland Bottoms <bottoms@debian.org>  Wed, 25 May 2016 17:19:57 -0400

rtl-sdr (0.5.3-8) unstable; urgency=high

  * Fix syntax errors for systemd-udevd in udev rules (Closes: #825073)

 -- A. Maitland Bottoms <bottoms@debian.org>  Tue, 24 May 2016 21:08:50 -0400

rtl-sdr (0.5.3-7) unstable; urgency=medium

  * better udev rules (more like camera devices in libgphoto2-6)

 -- A. Maitland Bottoms <bottoms@debian.org>  Tue, 10 May 2016 19:20:27 -0400

rtl-sdr (0.5.3-6) unstable; urgency=medium

  * Use ENV{ID_SOFTWARE_RADIO}=1 in udev rules (Closes: #823089)
  * Enable DETACH_KERNEL_DRIVER (Closes: 823022)
  * Make myself maintainer so I get the bug reports

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 08 May 2016 12:12:13 -0400

rtl-sdr (0.5.3-5) unstable; urgency=medium

  * Add watch fiule
  * place rtl-sdr -n comm section (Closes: #758077)
  * improve-librtlsdr-pc-file (Closes: #784912)
  * improve-scanning-range-parsing (LP: #1469478)

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 23 Aug 2015 10:35:42 -0400

rtl-sdr (0.5.3-4) unstable; urgency=low

  * Update to v0.5.3-12-ge3c03f7

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 08 Aug 2015 23:43:54 -0400

rtl-sdr (0.5.3-3) unstable; urgency=low

  * Update to v0.5.3-5-g6ee5573
    lib: handle events after canceling transfers
    lib: change default number of transfers to 15
    rtl_tcp: make all global variables static

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 13 Apr 2014 10:48:49 -0400

rtl-sdr (0.5.3-2) unstable; urgency=low

  * Upstream: lib: only print to stderr in tuner_r82xx_set_pll()
  * Update man pages (New -M (modulation) and -E (option) and ppm setting)
  * Have librtlsdr0 also install a blacklist for linux video drivers

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 08 Feb 2014 22:40:06 -0500

rtl-sdr (0.5.3-1) unstable; urgency=low

  * New upstream git tag release

 -- A. Maitland Bottoms <bottoms@debian.org>  Thu, 06 Feb 2014 20:45:38 -0500

rtl-sdr (0.5.2.7.3ab6-1~bpo70+1) wheezy-backports; urgency=low

  * Rebuild for wheezy-backports.

 -- A. Maitland Bottoms <bottoms@debian.org>  Tue, 21 Jan 2014 19:34:16 -0500

rtl-sdr (0.5.2.7.3ab6-1) unstable; urgency=low

  * New upstream snapshot

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 29 Dec 2013 21:37:19 -0500

rtl-sdr (0.5.1.14.360d-1~wheezy) stable; urgency=low

  * New upstream snapshot
  * GNU Radio LiveDVD 2013-1110

 -- A. Maitland Bottoms <bottoms@debian.org>  Mon, 11 Nov 2013 12:46:00 -0500

rtl-sdr (0.5.0.4.4914-2) unstable; urgency=low

  * Use kfreebsd libusb

 -- A. Maitland Bottoms <bottoms@debian.org>  Fri, 01 Nov 2013 17:16:42 -0400

rtl-sdr (0.5.0.4.4914-1) stable; urgency=low

  * New upstream snapshot (Closes: #701018).

src/rtl_sdr.c
/*
 * rtl-sdr, turns your Realtek RTL2832 based DVB dongle into a SDR receiver
 * Copyright (C) 2012 by Steve Markgraf <steve@steve-m.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <errno.h>
#include <signal.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef _WIN32
#include <unistd.h>
#else
#include <windows.h>
#include <io.h>
#include <fcntl.h>
#include "getopt/getopt.h"
#endif

#include "rtl-sdr.h"
#include "convenience/convenience.h"

#define DEFAULT_SAMPLE_RATE		2048000
#define DEFAULT_BUF_LENGTH		(16 * 16384)
#define MINIMAL_BUF_LENGTH		512
#define MAXIMAL_BUF_LENGTH		(256 * 16384)

static int do_exit = 0;
static uint32_t bytes_to_read = 0;
static rtlsdr_dev_t *dev = NULL;

void usage(void)
{
	fprintf(stderr,
		"rtl_sdr, an I/Q recorder for RTL2832 based DVB-T receivers\n\n"
		"Usage:\t -f frequency_to_tune_to [Hz]\n"
		"\t[-s samplerate (default: 2048000 Hz)]\n"
		"\t[-d device_index (default: 0)]\n"
		"\t[-g gain (default: 0 for auto)]\n"
		"\t[-p ppm_error (default: 0)]\n"
		"\t[-b output_block_size (default: 16 * 16384)]\n"
		"\t[-n number of samples to read (default: 0, infinite)]\n"
		"\t[-S force sync output (default: async)]\n"
		"\t[-D enable direct sampling (default: off)]\n"
		"\tfilename (a '-' dumps samples to stdout)\n\n");
	exit(1);
}

#ifdef _WIN32
BOOL WINAPI
sighandler(int signum)
{
	if (CTRL_C_EVENT == signum) {
		fprintf(stderr, "Signal caught, exiting!\n");
		do_exit = 1;
		rtlsdr_cancel_async(dev);
		return TRUE;
	}
	return FALSE;
}
#else
static void sighandler(int signum)
{
	signal(SIGPIPE, SIG_IGN);
	fprintf(stderr, "Signal caught, exiting!\n");
	do_exit = 1;
	rtlsdr_cancel_async(dev);
}
#endif

static void rtlsdr_callback(unsigned char *buf, uint32_t len, void *ctx)
{
	if (ctx) {
		if (do_exit)
			return;

		if ((bytes_to_read > 0) && (bytes_to_read < len)) {
			len = bytes_to_read;
			do_exit = 1;
			rtlsdr_cancel_async(dev);
		}

		if (fwrite(buf, 1, len, (FILE*)ctx) != len) {
			fprintf(stderr, "Short write, samples lost, exiting!\n");
			rtlsdr_cancel_async(dev);
		}

		if (bytes_to_read > 0)
			bytes_to_read -= len;
	}
}

int main(int argc, char **argv)
{
#ifndef _WIN32
	struct sigaction sigact;
#endif
	char *filename = NULL;
	int n_read;
	int r, opt;
	int gain = 0;
	int ppm_error = 0;
	int direct_sampling = 0;
	int sync_mode = 0;
	FILE *file;
	uint8_t *buffer;
	int dev_index = 0;
	int dev_given = 0;
	uint32_t frequency = 100000000;
	uint32_t samp_rate = DEFAULT_SAMPLE_RATE;
	uint32_t out_block_size = DEFAULT_BUF_LENGTH;

	while ((opt = getopt(argc, argv, "d:f:g:s:b:n:p:SD")) != -1) {
		switch (opt) {
		case 'd':
			dev_index = verbose_device_search(optarg);
			dev_given = 1;
			break;
		case 'f':
			frequency = (uint32_t)atofs(optarg);
			break;
		case 'g':
			gain = (int)(atof(optarg) * 10); /* tenths of a dB */
			break;
		case 's':
			samp_rate = (uint32_t)atofs(optarg);
			break;
		case 'p':
			ppm_error = atoi(optarg);
			break;
		case 'b':
			out_block_size = (uint32_t)atof(optarg);
			break;
		case 'n':
			bytes_to_read = (uint32_t)atof(optarg) * 2;
			break;
		case 'S':
			sync_mode = 1;
			break;
		case 'D':
			direct_sampling = 1;
			break;
		default:
			usage();
			break;
		}
	}

	if (argc <= optind) {
		usage();
	} else {
		filename = argv[optind];
	}

	if(out_block_size < MINIMAL_BUF_LENGTH ||
	   out_block_size > MAXIMAL_BUF_LENGTH ){
		fprintf(stderr,
			"Output block size wrong value, falling back to default\n");
		fprintf(stderr,
			"Minimal length: %u\n", MINIMAL_BUF_LENGTH);
		fprintf(stderr,
			"Maximal length: %u\n", MAXIMAL_BUF_LENGTH);
		out_block_size = DEFAULT_BUF_LENGTH;
	}

	buffer = malloc(out_block_size * sizeof(uint8_t));

	if (!dev_given) {
		dev_index = verbose_device_search("0");
	}

	if (dev_index < 0) {
		exit(1);
	}

	r = rtlsdr_open(&dev, (uint32_t)dev_index);
	if (r < 0) {
		fprintf(stderr, "Failed to open rtlsdr device #%d.\n", dev_index);
		exit(1);
	}
#ifndef _WIN32
	sigact.sa_handler = sighandler;
	sigemptyset(&sigact.sa_mask);
	sigact.sa_flags = 0;
	sigaction(SIGINT, &sigact, NULL);
	sigaction(SIGTERM, &sigact, NULL);
	sigaction(SIGQUIT, &sigact, NULL);
	sigaction(SIGPIPE, &sigact, NULL);
#else
	SetConsoleCtrlHandler( (PHANDLER_ROUTINE) sighandler, TRUE );
#endif

	/* Set direct sampling */
        if (direct_sampling)
                verbose_direct_sampling(dev, 2);

	/* Set the sample rate */
	verbose_set_sample_rate(dev, samp_rate);

	/* Set the frequency */
	verbose_set_frequency(dev, frequency);

	if (0 == gain) {
		 /* Enable automatic gain */
		verbose_auto_gain(dev);
	} else {
		/* Enable manual gain */
		gain = nearest_gain(dev, gain);
		verbose_gain_set(dev, gain);
	}

	verbose_ppm_set(dev, ppm_error);

	if(strcmp(filename, "-") == 0) { /* Write samples to stdout */
		file = stdout;
#ifdef _WIN32
		_setmode(_fileno(stdin), _O_BINARY);
#endif
	} else {
		file = fopen(filename, "wb");
		if (!file) {
			fprintf(stderr, "Failed to open %s\n", filename);
			goto out;
		}
	}

	/* Reset endpoint before we start reading from it (mandatory) */
	verbose_reset_buffer(dev);

	if (sync_mode) {
		fprintf(stderr, "Reading samples in sync mode...\n");
		while (!do_exit) {
			r = rtlsdr_read_sync(dev, buffer, out_block_size, &n_read);
			if (r < 0) {
				fprintf(stderr, "WARNING: sync read failed.\n");
				break;
			}

			if ((bytes_to_read > 0) && (bytes_to_read < (uint32_t)n_read)) {
				n_read = bytes_to_read;
				do_exit = 1;
			}

			if (fwrite(buffer, 1, n_read, file) != (size_t)n_read) {
				fprintf(stderr, "Short write, samples lost, exiting!\n");
				break;
			}

			if ((uint32_t)n_read < out_block_size) {
				fprintf(stderr, "Short read, samples lost, exiting!\n");
				break;
			}

			if (bytes_to_read > 0)
				bytes_to_read -= n_read;
		}
	} else {
		fprintf(stderr, "Reading samples in async mode...\n");
		r = rtlsdr_read_async(dev, rtlsdr_callback, (void *)file,
				      0, out_block_size);
	}

	if (do_exit)
		fprintf(stderr, "\nUser cancel, exiting...\n");
	else
		fprintf(stderr, "\nLibrary error %d, exiting...\n", r);

	if (file != stdout)
		fclose(file);

	rtlsdr_close(dev);
	free (buffer);
out:
	return r >= 0 ? r : -r;
}
